gtda/mapper/nerve.py

```diff
@@ -114,46 +133,101 @@ def fit_transform(self, X, y=None):
-        graph.vs["node_elements"] = node_elements
+        attribute_names = ["pullback_set_label", "partial_cluster_label",
+                           "node_elements"]
+        for i, node_attribute in enumerate(node_attributes):
+            graph.vs[attribute_names[i]] = node_attribute
 
         # Graph construction -- edges with weights given by intersection sizes
-        node_index_pairs, weights, intersections = \
-            self._generate_edge_data(node_elements)
+        node_index_pairs, weights, intersections, mapping = \
+            self._generate_edge_data(nodes)
         graph.es["weight"] = 1
         graph.add_edges(node_index_pairs)
         graph.es["weight"] = weights
         if self.store_edge_elements:
             graph.es["edge_elements"] = intersections
+        if self.contract_nodes:
+            fixed_points_mapping = _fixed_points(mapping)
+            graph.contract_vertices(
+                fixed_points_mapping, combine_attrs="first"
+                )
+            graph.delete_vertices(
+                [i for i in graph.vs.indices if i != fixed_points_mapping[i]]
+                )
 
         return graph
 
-    def _generate_edge_data(self, node_elements):
-        node_tuples = combinations(enumerate(node_elements), 2)
+    def _generate_edge_data(self, nodes):
+        node_tuples = combinations(enumerate(nodes), 2)
 
         node_index_pairs = []
         weights = []
         intersections = []
 
-        # Boilerplate is just to avoid boolean checking at each iteration
-        if not self.store_edge_elements:
-            for (node_1_idx, node_1_elements), (node_2_idx, node_2_elements) \
-                    in node_tuples:
-                intersection = np.intersect1d(node_1_elements, node_2_elements)
-                intersection_size = len(intersection)
-                if intersection_size >= self.min_intersection:
-                    node_index_pairs.append((node_1_idx, node_2_idx))
-                    weights.append(intersection_size)
+        if self.contract_nodes:
+            mapping = np.arange(len(nodes))
         else:
-            for (node_1_idx, node_1_elements), (node_2_idx, node_2_elements) \
-                    in node_tuples:
-                intersection = np.intersect1d(node_1_elements, node_2_elements)
-                intersection_size = len(intersection)
-                if intersection_size >= self.min_intersection:
-                    node_index_pairs.append((node_1_idx, node_2_idx))
-                    weights.append(intersection_size)
-                    intersections.append(intersection)
-
-        return node_index_pairs, weights, intersections
+            mapping = None
+
+        def _in_same_pullback_set(_node_tuple):
+            return _node_tuple[0][1][0] == _node_tuple[1][1][0]
+
+        def _do_nothing(*args):
+            pass
+
+        def _intersections_append(_intersection):
+            return intersections.append(_intersection)
```

Comment:
is it a good idea to have `intersections` defined with enclosing scope? (i.e. should `intersections` be a function arg?)

Reply:
I'm not aware of this as an anti-pattern to be honest, and I thought it was slightly more efficient. Given that these functions are not worthy of being taken outside the scope of `_generate_edge_data`, what is the issue?

Reply:
I mean, isn't this basically the idea of "closures"? https://realpython.com/inner-functions-what-are-they-good-for/#whats-a-closure

Reply:
good point about closures! that certainly applies to `_intersections_append()`, so my question was simply about the in-place change to the `intersections` list and whether one could have unexpected side effects (e.g. if `intersections` is modified elsewhere. a somewhat ugly alternative would be to promote `intersections` as a class attribute, but that's definitely overkill.

sorry for wasting your time 😬

Reply:
> sorry for wasting your time 😬
Oh please, we both love it :D